presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AddIntermediateAggregations.java

```diff
@@ -0,0 +1,191 @@
+        }
+
+        if (!(node instanceof ExchangeNode)) {
+            return Optional.empty();
+        }
+
+        ImmutableList.Builder<PlanNode> builder = ImmutableList.builder();
+        for (PlanNode source : node.getSources()) {
+            Optional<PlanNode> planNode = recurseThroughExchanges(lookup.resolve(source), lookup, idAllocator);
+            if (!planNode.isPresent()) {
+                return Optional.empty();
+            }
+            builder.add(planNode.get());
+        }
+        return Optional.of(ChildReplacer.replaceChildren(node, builder.build()));
+    }
+
+    private PlanNode addGatheringIntermediate(AggregationNode aggregation, PlanNodeIdAllocator idAllocator)
+    {
+        verify(aggregation.getGroupingKeys().isEmpty(), "Should be an un-grouped aggregation");
+        ExchangeNode gatheringExchange = ExchangeNode.gatheringExchange(idAllocator.getNextId(), ExchangeNode.Scope.LOCAL, aggregation);
+        return new AggregationNode(
+                idAllocator.getNextId(),
+                gatheringExchange,
+                outputsAsInputs(aggregation.getAssignments()),
+                aggregation.getGroupingSets(),
+                AggregationNode.Step.INTERMEDIATE,
+                aggregation.getHashSymbol(),
+                aggregation.getGroupIdSymbol());
+    }
+
+    /**
+     * Rewrite assignments so that inputs are in terms of the output symbols.
+     *
+     * Example:
+     * 'a' := sum('b') => 'a' := sum('a')
+     * 'a' := count(*) => 'a' := count('a')
+     */
+    private static Map<Symbol, AggregationNode.Aggregation> outputsAsInputs(Map<Symbol, AggregationNode.Aggregation> assignments)
+    {
+        ImmutableMap.Builder<Symbol, AggregationNode.Aggregation> builder = ImmutableMap.builder();
+        for (Map.Entry<Symbol, AggregationNode.Aggregation> entry : assignments.entrySet()) {
+            Symbol output = entry.getKey();
+            AggregationNode.Aggregation aggregation = entry.getValue();
+            builder.put(
+                    output,
+                    new AggregationNode.Aggregation(
+                            new FunctionCall(QualifiedName.of(aggregation.getSignature().getName()), ImmutableList.of(output.toSymbolReference())),
+                            aggregation.getSignature(),
+                            Optional.empty())); // No mask for INTERMEDIATE
+        }
+        return builder.build();
+    }
+
+    /**
+     * Rewrite assignments so that outputs are in terms of the input symbols.
+     *
+     * Example:
+     * 'a' := sum('b') => 'b' := sum('b')
+     */
+    private static Map<Symbol, AggregationNode.Aggregation> inputsAsOutputs(Map<Symbol, AggregationNode.Aggregation> assignments)
+    {
+        ImmutableMap.Builder<Symbol, AggregationNode.Aggregation> builder = ImmutableMap.builder();
+        for (Map.Entry<Symbol, AggregationNode.Aggregation> entry : assignments.entrySet()) {
+            // Should only have one input symbol
```

Comment:
good point, this works because the intermediates and outputs only work with one input and one output. I'll leave a comment about that.